### DISCUSSION TITLE
Autocompletion in source code for imported DLLs in (various) editor doesn't work.

### DISCUSSION THREAD

**Discussion by Geggon** [pythonnet/pythonnet#2042] [Q&A] [answered]
### Environment

-   Pythonnet version: v3.0.1
-   Python version: Currently 3.8 (Also tested on other versions)
-   Operating System: Win 11
-   .NET Runtime: 4.8.09032

### Details

Hi everyone, I hope you can help me with the following problem. I have already found diverese posts here and on other sites, but unfortunately couldn't solve the problem. The goal is to read a DLL, which is normally used in C#, into Python. There is a large list of methods, which is greatly simplified by the autocompletion. Unfortunately, I lose a strong convenience in the Python editor, that the interpreter in the static code can't execute the autocompletion (my guess). Thus, I cannot see in the editor what commands are available to me. 

BUT: If I execute the same commands directly in the console, the autocompletion works perfectly. Also all methods are loaded, which are in the modules.

I have already tried this with different IDEs: PyCharm, VS2019/22, VSC, Eclipse. I have also had unsuccessful attempts with IronPython.

Is there any way that this DLL gets pre-read in the editor or that the contents are known. I have read in some forum posts that so-called stubs should help... Unfortunately, however, I failed in the implementation.

The DLL is obtained directly from the project website: [https://www.nuget.org/packages/Aml.Engine](url) [Download -> AMLEngine 2.1 v.1.7 - In this are also different .NET versions, which I have tested]. The DLLs are in the code directory and also added it to the system environment variables for testing. I also thought that replacing Import * with the names e.g. CAEXDocument would help.... sadly not.

```python
import clr

clr.AddReference("DLLs/AML.Engine")

import Aml.Engine
import Aml.Engine.CAEX
from Aml.Engine.CAEX import AttributeFamilyType
from Aml.Engine.CAEX import CAEXDocument

# from Aml.Engine.CAEX import *
# from Aml.Engine.CAEX.Extensions import *

def main():
    doc = CAEXDocument.New_CAEXDocument()
    ih = doc.CAEXFile.InstanceHierarchy.Append("IH1")
    # etc. here goes the source code

if __name__ == '__main__':
    main()
```
The autocomplete should actually suggest _NEW_CAEXDocument_ to me after _doc=CAEXDocument._ or other methods which follow. Unfortunately I don't get any of them.

The same also by _ih=doc.---_

I get error messages like: 

```
- Unresolved reference 'CAEXDocument'
- Cannot find reference 'AddReference' in 'clr.py'
```
However, the script can be executed without problems.

Many suggestions in forum posts were (so I feel) a bit far from the topic or my knowledge is not sufficient for this, since Python is quite new to me.

I would be very grateful for any possible solutions. If there are any uncertainties or questions, I will answer them as fast as possible.

**Comment by gunhasiz** [comment]
I have the same problem, unfortunately found no solution. Have you?

**Reply by Geggon** [reply]
Unfortunately, I still have the problem and couldn't solve it until now. In order to be able to use the code in Python, I write it in C# and then execute it in Python. Of course, a few adjustments have to be made. For this the console helps immensely, because there the suggestions (autocompletion) are made once the modules from the DLLs have been loaded.

**Comment by MHDante** [comment]
In order to get autocomplete, you'll likely need to generate stubs for the c# dlls.

Afaik, there's 2 projects that can do this:

https://github.com/gtalarico/ironpython-stub
https://github.com/MHDante/pythonnet-stub-generator

You can then use the stubs as indicated [in your IDE](https://www.jetbrains.com/help/pycharm/stubs.html)

**Reply by Geggon** [reply]
Thank you for your answer. I had actually tried it once via the Ironpython stubs... but unfortunately without success. But I don't know the second link "pythonnet-stub-generator" yet. I'll test it in the next few days and report back.

**Reply by Geggon** [reply]
Thank you very much. I could generate and integrate the stubs with the second link. This means that I can now write AML files in Python.

**Reply by Geggon** [reply]
If I integrate the stubs as you have sent them to me, I can unfortunately no longer use the imported modules (import Aml.Engine etc.), as these are practically overwritten by the stubs (I guess). And I get an error message: NameError: name 'CAEXDocument' is not defined

However, if I remove the stubs, I can execute the code correctly, but the autocompletion no longer works. 

Is there any other way to read the information from the __init__.pyi data without overwriting my modules?

The class exists in the __init__.pyi file.

![grafik](https://github.com/pythonnet/pythonnet/assets/120031368/0c786795-490e-4314-8ead-42060f80f037)

![grafik](https://github.com/pythonnet/pythonnet/assets/120031368/d9f25d1a-e7e9-4e2e-b842-5a6fb92307bf)

**Reply by MHDante** [reply, marked as answer]
The pyi files should not go in your source code. Python doesn't need to know about the pyi files, they are only there for the use to use. 

If you follow the instructions here: 
https://www.jetbrains.com/help/pycharm/stubs.html#install-stub

The stubs should not change anything about the runtime behaviour of the software.

**Reply by Geggon** [reply]
Thank you very much for your support. I had problems integrating the stubs at the beginning. Now it works. I am currently using the stubs locally. I will upload them to PyPi and test it.
